pkg/local_object_storage/pilorama/boltdb.go

```diff
@@ -280,15 +298,62 @@ func (t *boltForest) TreeApply(d CIDDescriptor, treeID string, m *Move, backgrou
 		}
 	}
 
-	return t.db.Batch(func(tx *bbolt.Tx) error {
-		bLog, bTree, err := t.getTreeBuckets(tx, d.CID, treeID)
-		if err != nil {
-			return err
-		}
+	if t.db.MaxBatchSize == 1 {
+		return t.db.Update(func(tx *bbolt.Tx) error {
+			bLog, bTree, err := t.getTreeBuckets(tx, d.CID, treeID)
+			if err != nil {
+				return err
+			}
 
-		lm := &LogMove{Move: *m}
-		return t.applyOperation(bLog, bTree, lm)
-	})
+			var lm LogMove
+			return t.applyOperation(bLog, bTree, []*Move{m}, &lm)
+		})
+	}
+
+	ch := make(chan error, 1)
+	t.addBatch(d, treeID, m, ch)
+	return <-ch
+}
+
+func (t *boltForest) addBatch(d CIDDescriptor, treeID string, m *Move, ch chan error) {
+	t.mtx.Lock()
+	for i := 0; i < len(t.batches); i++ {
+		t.batches[i].mtx.Lock()
+		if t.batches[i].timer == nil {
+			t.batches[i].mtx.Unlock()
+			copy(t.batches[i:], t.batches[i+1:])
+			t.batches = t.batches[:len(t.batches)-1]
+			i--
+			continue
+		}
+		valid := t.batches[i].cid.Equals(d.CID) && t.batches[i].treeID == treeID
+		if valid {
+			t.batches[i].results = append(t.batches[i].results, ch)
+			t.batches[i].operations = append(t.batches[i].operations, m)
+			if len(t.batches[i].operations) == t.db.MaxBatchSize {
+				t.batches[i].timer = nil
+				t.batches[i].mtx.Unlock()
+				b := t.batches[i]
+				t.mtx.Unlock()
+				b.trigger()
+				return
+			}
+			t.batches[i].mtx.Unlock()
+			t.mtx.Unlock()
+			return
+		}
+		t.batches[i].mtx.Unlock()
+	}
+	b := &batch{
+		db:         t,
+		cid:        d.CID,
+		treeID:     treeID,
+		results:    []chan<- error{ch},
+		operations: []*Move{m},
+	}
+	b.timer = time.AfterFunc(time.Millisecond*20, b.trigger)
```

Comment:
does 20ms _really_ help? have you tried other values?

Reply:
What's the expected number of operations per second and `MaxBatchSize`?

Reply:
What do you mean by `20ms really help`?
The purpose of this PR is to batch operations _over a single container_, this is what helps.
Actually, I don't remember where 20ms came from (10ms is default in bbolt). Anyway, we should have `db.MaxBatchDelay` here.

Fixed

Reply:
> What do you mean by 20ms really help?

i meant that i can say nothing about that duration (is it short or long) so i am interested if you tried any other values and if you measured how different values affect performance, if yes, i would add a constant with a comment or make that configurable (`db.MaxBatchDelay` is also ok for me). in other words, "What's the expected number of operations per second?" as @roman-khimov asked.

Reply:
We have no expected number of operations, because this should work in any circumstances. I think bolt default is fine here. I am also thinking about making this setting different from the "normal batch size/delay", but we need tests for this, will do in another PR if it helps.

Reply:
You have to have some expectations, this service has some performance limits whether you like it or not. BoltDB itself is not infinite wrt to how many reads/writes per second it can deliver. If you know your limits and expectations, you can tune parameters like this better (or evaluate other approaches).